### DISCUSSION TITLE
Is it possible to narrow from `object` to `dict[str, object]` under strict mode?

### DISCUSSION THREAD

**Discussion by ghost** [microsoft/pyright#6577] [Q&A]
I'm running under strict mode (which enables the reportUnknownArgumentType diagnostic rule). I have an `object` that's come from `json.loads` and I want to assert that it's more specifically a `dict[str, object]`. With non-compound types (e.g. `int`) this would be easy: `assert isinstance(o, int)`, but I haven't found a way to assert the type of this dict's keys without running into `reportUnknownVariableType` errors. I also tried using a TypeGuard to assert this type but again couldn't write the code without unknown types.

My question as code:

```python
def f() -> dict[str, object]:
   result: object = json.loads('{"one": 1}')
   assert isinstance(result, dict), result   # now we have type dict[Unknown, Unknown]
   
   assert ...  # what goes here?
   
   return result
```

**Comment by erictraut** [comment]
Here's one approach:

```python
type JSON = None | str | float | dict[str, JSON] | list[JSON]

def f() -> dict[str, JSON]:
    result: JSON = json.loads('{"one": 1}')
    assert isinstance(result, dict), result
    return result
```

**Reply by ghost** [reply]
Your example transitions the value from type `Any` to `JSON`, but in my situation I start with `object`.

My question more precisely is whether there's any type-assertion that can narrow `object` to `dict[str, object]`.

**Reply by erictraut** [reply]
The `isinstance` call doesn't allow you specify a specialized type like `dict[str, object]`, so you would need to use some other technique to avoid the `Unknown`. A user-defined type guard (a'la PEP 647) would be one option. Another option, if you're really starting with a value whose type is simply `object`, is to use a `cast` to cast it to a known `dict` type like `dict[str, Any]`.

**Reply by ghost** [reply]
Thanks. I ran across this problem again, in a slightly different form. Do I need to use cast here? Or is it possible to write this as a typeguard that doesn't need cast (under strict pyright settings)?

```python
type JSON = None | str | float | dict[str, JSON] | list[JSON]

def _assert_dict_of_strings(json: JSON) -> dict[str, str]:
    assert isinstance(json, dict), json
    for key, val in json.items():
        assert isinstance(key, str), (key, json)
        assert isinstance(val, str), (val, json)
    # https://github.com/microsoft/pyright/discussions/6577
    return cast(dict[str, str], json)
```

**Reply by erictraut** [reply]
Yeah, you'd need to create a user-defined type guard here. And that type guard would be "unsafe" in that it's violating the type system because it would be claiming that the type of the object is a `dict[str, str]` when it's actually `dict[str, JSON]`. If you want to make this truly type safe, you'd need to create a new object.

```python
def _get_dict_of_strings(json: dict[str, JSON]) -> dict[str, str]:
    return {key: val for key, val in json.items() if isinstance(val, str)}
```

**Reply by ghost** [reply]
I considered that solution, but in the cases where it removes (non-str) values, I'd rather have an AssertionError.
I think, as you said, I'll need a custom typeguard that has a `#pylint:ignore` inside it.
That's was my conclusion before I asked, but I wanted to be more sure.
Thanks.

Any chance you'd help me write the mentioned custom typeguard?

Currently I'm using this:

```python
def assert_dict_of_strings(json: object) -> dict[str, str]:
    assert isinstance(json, dict), json
    # https://github.com/microsoft/pyright/discussions/6577
    for key, val in json.items():  # pyright:ignore[reportUnknownVariableType]
        assert isinstance(key, str), (key, json)
        assert isinstance(val, str), (val, json)
    return cast(dict[str, str], json)
```

Actually, is a TypeGuard preferable to this? I think my co-maintainers will find this (relatively) simple to understand whereas TypeGuard is a new concept they haven't encountered before.
